### ISSUE TITLE
Systemd user services are not stopped automatically

### ISSUE THREAD

**Issue by Gerschtli** [NixOS/nixpkgs#65462]
**Describe the bug**
After disabling a systemd user service, the service file got removed but the service is not stopped and keeps running.

If you have any pointers on how to fix this, I would love to create a PR!

**To Reproduce**
1. Add to `configuration.nix`:
```nix
{
  programs.nm-applet.enable = true;
}
```
2. Run `nixos-rebuild test`
3. User service runs perfectly fine
4. Remove the line added in 1.
5. Run `nixos-rebuild test`
6. Service is still running, but unit file is not found:
```
$> systemctl --user status nm-applet.service
● nm-applet.service
   Loaded: not-found (Reason: Unit nm-applet.service not found.)
   Active: active (running) since Sat 2019-07-27 11:24:01 CEST; 14min ago
 Main PID: 1085 (.nm-applet-wrap)
   CGroup: /user.slice/user-1000.slice/user@1000.service/nm-applet.service
           └─1085 /nix/store/r449f15ff5wa2lparj3hspdbvxn98nq7-network-manager-applet-1.8.20/bin/nm-applet

Jul 27 11:24:01 argon systemd[1024]: Started Network manager applet.
Jul 27 11:24:01 argon .nm-applet-wrap[1085]: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
Jul 27 11:34:13 argon systemd[1024]: nm-applet.service: Current command vanished from the unit file, execution of the command list won't be resumed.
```

**Expected behavior**
After I disable a systemd user service I expect that the unit file gets deleted (which is currently working) and that the service get stopped.

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.59, NixOS, 19.03.173144.973a2705351 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(tobias): `"home-manager"`
 - channels(root): `"nixos-19.03.173144.973a2705351"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
  - programs.nm-applet
  - programs.xss-lock
  - services.unclutter-xfixes
```
There are probably more module affected by this problem, but these are the ones I have tested.

**Comment by Ma27** [comment]
> There are probably more module affected by this problem, but these are the ones I have tested.

In fact, every module which uses `systemd.user.services.<service-name>` is affected by this.

> If you have any pointers on how to fix this, I would love to create a PR!

I had a look at this about a year ago, but didn't manage to create a proper fix. You may want to have a look at `switch-to-configuration.pl` (`./nixos/modules/system/activation/switch-to-configuration.pl`) which activates a new nixos configuration and deals with changed systemd units (e.g. reloads/restarts modified ones and starts newly added units).

On a recent master (f217c3b6fe2066518ac1f0429d3bb2fe1346e5c3) all system-wide units are queried in this script in line 71 by loading those using dbus (`Net::DBus->system->get_service("org.freedesktop.systemd1")->get_object("/org/freedesktop/systemd1");`).

It seems as systemd stores and manages user units somehow differently, but I didn't manage to find a proper way to load those for each active user as well..

I'm pretty sure that this is the biggest issue, the reload/restart/whatever logic in this script should be reusable for user units as well. The only thing to keep in mind is that it's probably harmful to restart *all* changed user units (as you don't want X to restart after most nixos configuration changes).

Does this help a little bit?

**Comment by Gerschtli** [comment]
My capabilites in perl and my deep knowledge in systemd and dbus make it more difficult to provide a fix.

In https://github.com/NixOS/nixpkgs/pull/44990 the logic for reloading user units was introduced. Maybe anyone of the reviewers or the author have some interesting input.

CC @Ma27 @Mic92 @edolstra

**Comment by Ma27** [comment]
Yeah, I actually wrote this, but stopped at the point where I wanted to implement start/stop behavior. Unfortunately I don't have sufficient time ATM to take a closer look again :)

**Comment by Gerschtli** [comment]
ASAICT this is still an issue.

**Comment by flokli** [comment]
Cross-referencing: https://github.com/NixOS/nixpkgs/issues/24409

**Comment by flokli** [comment]
still relevant.

**Comment by flokli** [comment]
still relevant.
